ocamlc -c and ocamlopt -c differ for C files

**Original bug ID:** 7678
**Reporter:** @dbuenzli
**Status:** acknowledged (set by @xavierleroy on 2017-11-26T16:21:45Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**OS:** Linux
**Version:** 4.06.0
**Target version:** 4.07.0+dev/beta2/rc1/rc2
**Category:** compiler driver
**Monitored by:** @gasche @yallop @dra27 @dbuenzli

## Bug description

I don't know if that is intended but it seems that `-fPIC` is not passed in the second case. 

This makes `ocamlmklib` to fail to produce libraries for C object files generated with ocamlopt -c.

## Steps to reproduce

# Our C file 
> cat a.c
#include <stdlib.h>
char *get_editor (void) { return getenv("EDITOR"); }

# Here all is well
> ocamlc -verbose -c a.c 
+ gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT   -c    -I'/home/dbuenzli/.opam/default/lib/ocaml' 'a.c'
> ocamlmklib a.o

# Here we fail note the missing `-fPIC` w.r.t above invocation
> ocamlopt -verbose -c a.c 
+ gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp -D_FILE_OFFSET_BITS=64 -D_REENTRANT   -c    -I'/home/dbuenzli/.opam/default/lib/ocaml' 'a.c'

## Additional information

> cat /etc/system-release
CentOS Linux release 7.4.1708 (Core) 

> ocamlc -config 
 ocamlc -config 
version: 4.06.0
standard_library_default: /home/dbuenzli/.opam/default/lib/ocaml
standard_library: /home/dbuenzli/.opam/default/lib/ocaml
standard_runtime: /home/dbuenzli/.opam/default/bin/ocamlrun
ccomp_type: cc
c_compiler: gcc
ocamlc_cflags: -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp -fPIC
ocamlc_cppflags: -D_FILE_OFFSET_BITS=64 -D_REENTRANT
ocamlopt_cflags: -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp
ocamlopt_cppflags: -D_FILE_OFFSET_BITS=64 -D_REENTRANT
bytecomp_c_compiler: gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT
native_c_compiler: gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp -D_FILE_OFFSET_BITS=64 -D_REENTRANT
bytecomp_c_libraries: -lm  -ldl -lcurses -lpthread                  
native_c_libraries: -lm  -ldl
native_pack_linker: ld -r -o 
ranlib: ranlib
cc_profile: -pg
architecture: amd64
model: default
int_size: 63
word_size: 64
system: linux
asm: as
asm_cfi_supported: true
with_frame_pointers: false
ext_exe: 
ext_obj: .o
ext_asm: .s
ext_lib: .a
ext_dll: .so
os_type: Unix
default_executable_name: a.out
systhread_supported: true
host: x86_64-unknown-linux-gnu
target: x86_64-unknown-linux-gnu
profiling: true
flambda: false
spacetime: false
safe_string: false
default_safe_string: true
flat_float_array: true
afl_instrument: false
windows_unicode: false
exec_magic_number: Caml1999X011
cmi_magic_number: Caml1999I022
cmo_magic_number: Caml1999O022
cma_magic_number: Caml1999A022
cmx_magic_number: Caml1999Y022
cmxa_magic_number: Caml1999Z022
ast_impl_magic_number: Caml1999M022
ast_intf_magic_number: Caml1999N022
cmxs_magic_number: Caml1999D022
cmt_magic_number: Caml1999T022

**Comment author:** @dbuenzli

Forgot to add the failing `ocamlmklib` but you get the idea:

> ocamlmklib a.o
/usr/bin/ld: a.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
a.o: error adding symbols: Bad value
collect2: error: ld returned 1 exit status

--- a/Changes
+++ b/Changes
@@ -209,6 +209,11 @@ Working version
   building shared libraries like `-output-obj`.
   (Florian Angeletti, review by Nicolás Ojeda Bär)
 
+* #7678, #8631: ocamlc -c and ocamlopt -c pass same switches to the C
+  compiler when compiling .c files (in particular, this means ocamlopt
+  passes -fPIC on systems requiring it for shared library support).
+  (David Allsopp, report by Daniel Bünzli, review by Sébastien Hinderer)
+
 - #9074: reworded error message for non-regular structural types
   (Florian Angeletti, review by Jacques Garrigue and Leo White,
    report by Chas Emerick)

--- a/Makefile.config.in
+++ b/Makefile.config.in
@@ -193,8 +193,10 @@ NATDYNLINK=@natdynlink@
 NATDYNLINKOPTS=@natdynlinkopts@
 SYSLIB=@syslib@
 MKLIB=@mklib@
-OCAMLOPT_CFLAGS=@ocamlopt_cflags@
-OCAMLOPT_CPPFLAGS=@ocamlopt_cppflags@
+# #7678: ocamlopt uses these only to compile .c files, and the behaviour for the
+#        two drivers should be identical.
+OCAMLOPT_CFLAGS=@ocamlc_cflags@
+OCAMLOPT_CPPFLAGS=@ocamlc_cppflags@
 NATIVECCLIBS=@nativecclibs@
 SYSTHREAD_SUPPORT=@systhread_support@
 PACKLD=@PACKLD@

--- a/configure
+++ b/configure
@@ -804,8 +804,6 @@ AR
 shebangscripts
 long_shebang
 iflexdir
-ocamlopt_cppflags
-ocamlopt_cflags
 ocamlc_cppflags
 ocamlc_cflags
 nativecclibs
@@ -2748,8 +2746,6 @@ internal_cflags=""
 internal_cppflags=""
 ocamlc_cflags=""
 ocamlc_cppflags=""
-ocamlopt_cflags=""
-ocamlopt_cppflags=""
 oc_ldflags=""
 with_sharedlibs=true
 ostype="Unix"
@@ -2806,8 +2802,6 @@ VERSION=4.11.0+dev0-2019-10-18
 # Note: This is present for the flexdll bootstrap where it exposed as the old
 # TOOLPREF variable. It would be better if flexdll where updated to require
 # WINDRES instead.
-
-
 
 
 
@@ -16886,8 +16880,6 @@ oc_cflags="$common_cflags $internal_cflags"
 oc_cppflags="$common_cppflags $internal_cppflags"
 ocamlc_cflags="$common_cflags $sharedlib_cflags"
 ocamlc_cppflags="$common_cppflags"
-ocamlopt_cflags="$common_cflags"
-ocamlopt_cppflags="$common_cppflags"
 cclibs="$cclibs $mathlib"
 
 case $host in #(

--- a/configure.ac
+++ b/configure.ac
@@ -47,8 +47,6 @@ internal_cflags=""
 internal_cppflags=""
 ocamlc_cflags=""
 ocamlc_cppflags=""
-ocamlopt_cflags=""
-ocamlopt_cppflags=""
 oc_ldflags=""
 with_sharedlibs=true
 ostype="Unix"
@@ -108,8 +106,6 @@ AC_SUBST([bytecclibs])
 AC_SUBST([nativecclibs])
 AC_SUBST([ocamlc_cflags])
 AC_SUBST([ocamlc_cppflags])
-AC_SUBST([ocamlopt_cflags])
-AC_SUBST([ocamlopt_cppflags])
 AC_SUBST([iflexdir])
 AC_SUBST([long_shebang])
 AC_SUBST([shebangscripts])
@@ -1788,8 +1784,6 @@ oc_cflags="$common_cflags $internal_cflags"
 oc_cppflags="$common_cppflags $internal_cppflags"
 ocamlc_cflags="$common_cflags $sharedlib_cflags"
 ocamlc_cppflags="$common_cppflags"
-ocamlopt_cflags="$common_cflags"
-ocamlopt_cppflags="$common_cppflags"
 cclibs="$cclibs $mathlib"
 
 AS_CASE([$host],

--- a/utils/ccomp.ml
+++ b/utils/ccomp.ml
@@ -89,11 +89,12 @@ let compile_file ?output ?(opt="") ?stable_name name =
          (match !Clflags.c_compiler with
           | Some cc -> cc
           | None ->
-              let (cflags, cppflags) =
-                  if !Clflags.native_code
-                  then (Config.ocamlopt_cflags, Config.ocamlopt_cppflags)
-                  else (Config.ocamlc_cflags, Config.ocamlc_cppflags) in
-              (String.concat " " [Config.c_compiler; cflags; cppflags]))
+              (* #7678: ocamlopt only calls the C compiler to process .c files
+                 from the command line, and the behaviour between
+                 ocamlc/ocamlopt should be identical. *)
+              (String.concat " " [Config.c_compiler;
+                                  Config.ocamlc_cflags;
+                                  Config.ocamlc_cppflags]))
          debug_prefix_map
          (match output with
           | None -> ""

--- a/utils/config.mli
+++ b/utils/config.mli
@@ -51,10 +51,14 @@ val ocamlc_cppflags : string
 (** The flags ocamlc should pass to the C preprocessor *)
 
 val ocamlopt_cflags : string
-(** The flags ocamlopt should pass to the C compiler *)
+  [@@ocaml.deprecated "Use ocamlc_cflags instead."]
+(** @deprecated {!ocamlc_cflags} should be used instead.
+    The flags ocamlopt should pass to the C compiler *)
 
 val ocamlopt_cppflags : string
-(** The flags ocamlopt should pass to the C preprocessor *)
+  [@@ocaml.deprecated "Use ocamlc_cppflags instead."]
+(** @deprecated {!ocamlc_cppflags} should be used instead.
+    The flags ocamlopt should pass to the C preprocessor *)
 
 val bytecomp_c_libraries: string
 (** The C libraries to link with custom runtimes *)

--- a/utils/config.mlp
+++ b/utils/config.mlp
@@ -35,7 +35,9 @@ let c_has_debug_prefix_map = %%CC_HAS_DEBUG_PREFIX_MAP%%
 let as_has_debug_prefix_map = %%AS_HAS_DEBUG_PREFIX_MAP%%
 let ocamlc_cflags = "%%OCAMLC_CFLAGS%%"
 let ocamlc_cppflags = "%%OCAMLC_CPPFLAGS%%"
-let ocamlopt_cflags = "%%OCAMLOPT_CFLAGS%%"
+(* #7678: ocamlopt uses these only to compile .c files, and the behaviour for
+          the two drivers should be identical. *)
+let ocamlopt_cflags = "%%OCAMLC_CFLAGS%%"
 let ocamlopt_cppflags = "%%OCAMLOPT_CPPFLAGS%%"
 let bytecomp_c_libraries = "%%BYTECCLIBS%%"
 (* bytecomp_c_compiler and native_c_compiler have been supported for a

